packages/lib/src/interactions/box.ts

```diff
@@ -0,0 +1,61 @@
+import { Box3, Vector3 } from 'three';
+
+import type { Size } from '../vis/models';
+
+const ZERO_VECTOR = new Vector3(0, 0, 0);
+
+class Box extends Box3 {
+  public get size(): Size {
+    const { x: width, y: height } = this.getSize(new Vector3());
+    return { width, height };
+  }
+
+  public get center(): Vector3 {
+    return this.getCenter(new Vector3());
+  }
+
+  public static empty(center = ZERO_VECTOR) {
+    return new Box(center.clone(), center.clone());
+  }
+
+  public static fromPoints(...points: Vector3[]) {
+    return new Box().setFromPoints(points);
+  }
+
+  public expandBySize(width: number, height: number): this {
+    return this.expandByVector(new Vector3(width, height, 0).divideScalar(2));
+  }
+
+  public expandToRatio(ratio: number | undefined): this {
+    if (ratio === undefined) {
+      return this;
+    }
+
+    const { width, height } = this.size;
+    const originalRatio = width / height;
+
+    return this.expandBySize(
+      originalRatio < ratio ? height * ratio - width : 0,
+      originalRatio < ratio ? 0 : width / ratio - height
+    );
+  }
+
+  public keepWithin(area: Box): this {
+    const { center, size } = this;
+    const { width: areaWidth, height: areaHeight } = area.size;
+
+    const centerClampingBox = Box.empty(area.center).expandBySize(
+      Math.max(areaWidth - size.width, 0),
+      Math.max(areaHeight - size.height, 0)
+    );
+
+    const shift = centerClampingBox
+      .clampPoint(center, new Vector3())
+      .sub(center)
+      .setZ(0); // cancel `z` shift
```

Comment:
The downside of working with `Box3` and `Vector3`: we end up having to correct the `z` component in a few places so the camera keeps `z=5` and doesn't move behind the scene... Perhaps we can revisit this in the future if we think of something clever.